### DISCUSSION TITLE
Why `from __future__ import annotations` is needed for correct OpenAPI ?

### DISCUSSION THREAD

**Discussion by jc-louis** [fastapi/fastapi#12415] [Questions] [answered]
```python
from __future__ import annotations

from fastapi import FastAPI, __version__

app = FastAPI()


@app.get("/test")
def test_call() -> None:
    return None

openapi = app.openapi()

print("FastAPI version", __version__)
print(openapi["paths"]["/test"]["get"]["responses"]["200"]["content"]["application/json"]["schema"])
```

- With `from __future__ import annotations` I get `{'type': 'null', 'title': 'Response Test Call Test Get'}` 
- Without `from __future__ import annotations` I get `{}`

### Operating System

macOS

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

Python 3.12.3

**Comment by Reemyos** [comment]
It seems like this problem originates here:
```Python
# File: fastapi/dependencies/utils.py
def get_typed_annotation(annotation: Any, globalns: Dict[str, Any]) -> Any:
    if isinstance(annotation, str):
        annotation = ForwardRef(annotation)
        annotation = evaluate_forwardref(annotation, globalns, globalns)
    return annotation
```
Where:
- With `from __future__ import annotations` the annotation arrives as a str ('None') and then the function returns `<Class NonType>`
- Without `from __future__ import annotations` the annotation arrives as `None` and the function returns `None`

I'm still quite new to this repository so I'm not suggesting a fix at the moment but I hope this helps you or someone else find a solution.

**Comment by sinisaos** [comment, marked as answer]
@jc-louis There is a great explanation in [this discussion](https://github.com/fastapi/fastapi/discussions/9723#discussioncomment-6266695), but the fix is ​​not good. We should add one more check to [this method](https://github.com/fastapi/fastapi/blob/13a18f80b3b0c7f0272c67b6ac7b73aafbb8584c/fastapi/dependencies/utils.py#L246). It should be something like this

```python
def get_typed_return_annotation(call: Callable[..., Any]) -> Any:
    signature = inspect.signature(call)
    annotation = signature.return_annotation

    # for python>=3.10, module types is returned but we cannot use in Python 3.8 and 3.9
    # from types import NoneType
    # if isinstance(annotation, NoneType):
    #     return str(None)
    
    # if annotaion is "None" return "None" not None
    if str(annotation) == str(None):
        return str(None)

    if annotation is inspect.Signature.empty:
        return None

    globalns = getattr(call, "__globals__", {})
    return get_typed_annotation(annotation, globalns)
```
Hope that helps.

**Comment by Reemyos** [comment]
I also found that if you do this:
```Python
from typing import Annotated

from fastapi import FastAPI, __version__

app = FastAPI()


@app.get("/test")
def test_call() -> Annotated[None, "This is a test call"]:
    return None

openapi = app.openapi()

print("FastAPI version", __version__)
print(openapi["paths"]["/test"]["get"]["responses"]["200"]["content"]["application/json"]["schema"])
```
It will give you your desired result.

**Reply by sinisaos** [reply]
@Reemyos It's even better because with `Annotated` we don't have to change the source code. Great.
